ompi/mpi/c/unpack.c

```diff
@@ -79,12 +79,64 @@ int MPI_Unpack(const void *inbuf, int insize, int *position,
 
     OPAL_CR_ENTER_LIBRARY();
 
+/*
+ * If a datatype's description contains a single element that describes
+ * a large vector that path is reasonably optimized in pack/unpack. On
+ * the other hand if the count and datatype combined describe the same
+ * vector that is processed one element at a time.
+ *
+ * So at the top level we morph the call if the count and datatype look
+ * like a good vector.
+ *
+ * Note the FLAG_CONTIG only means the data within the datatype is
+ * contiguous, not necessarily that a count,dt pair is still contiguous.
+ * Eg a resized datatype like [int4b,empty4b] has the contiguous flag.
+ * We want to activate the datatype creation for both types that are
+ * non-contiguous individually or when taken as a count.
+ */
+    bool created_larger_datatype = false;
+    MPI_Datatype larger_datatype;
+
+    /*
+     *  This magic number 250 is just empirical from one datapoint on one
+     *  machine, using [int,space] as the datatype.  There's some overhead
+     *  in constructing a new datatype, so just walking a sequence of
+     *  description elements is better for a short list of elements vs
+     *  creating a potentially shorter list and hoping the vector-walking
+     *  of the new elements is faster.  This could maybe be tuned dynamically
+     *  but it doesn't really seem worth it.
+     */
+    if (outcount > 250) {
```

Comment:
This code is identical to the pack case.  Maybe an inline function that is shared between the two?